src/server/pfs/server/driver_v2.go

```diff
@@ -168,39 +165,87 @@ func (d *driverV2) withFileSet(ctx context.Context, repo, commit string, f func(
-		return fr.Get(w, true)
-	}); err != nil {
+func (d *driverV2) getTar(ctx context.Context, commit *pfs.Commit, glob string, w io.Writer) error {
+	compactedPaths := []string{compactedCommitPath(commit)}
+	prefix := globLiteralPrefix(glob)
+	mf, err := matchFunc(glob)
+	if err != nil {
 		return err
 	}
-	// Close a tar writer to create tar EOF padding.
-	return tar.NewWriter(w).Close()
+	mr, err := d.storage.NewMergeReader(ctx, compactedPaths, index.WithPrefix(prefix))
+	if err != nil {
+		return err
+	}
+	filter := &fileset.IndexFilter{
+		R: mr,
+		F: func(idx *index.Index) bool {
+			return mf(idx.Path)
+		},
+	}
+	// TODO: remove absolute paths on the way out?
+	// nonAbsolute := &fileset.HeaderMapper{
+	// 	R: filter,
+	// 	F: func(th *tar.Header) *tar.Header {
+	// 		th.Name = "." + th.Name
+	// 		return th
+	// 	},
+	// }
+	return fileset.WriteTarStream(w, filter)
 }
 
-func (d *driverV2) listFileV2(pachClient *client.APIClient, file *pfs.File, full bool, history int64, f func(*pfs.FileInfoV2) error) error {
+func (d *driverV2) listFileV2(pachClient *client.APIClient, file *pfs.File, full bool, history int64, cb func(*pfs.FileInfoV2) error) error {
 	ctx := pachClient.Ctx()
 	if err := validateFile(file); err != nil {
 		return err
 	}
 	if err := d.checkIsAuthorized(pachClient, file.Commit.Repo, auth.Scope_READER); err != nil {
 		return err
 	}
-	// TODO: handle children of a directory
-	return d.getTarConditional(ctx, file.Commit.Repo.Name, file.Commit.ID, file.Path, func(fr *FileReader) error {
-		return f(fr.Info())
-	})
-}
-
-var globRegex = regexp.MustCompile(`[*?[\]{}!()@+^]`)
-
-func globLiteralPrefix(glob string) string {
-	idx := globRegex.FindStringIndex(glob)
-	if idx == nil {
-		return glob
+	name := strings.TrimRight(file.Path, "/")
+	// exact match
+	getExact := func() (fileset.FileSource, error) {
+		mr, err := d.storage.NewMergeReader(ctx, []string{compactedCommitPath(file.Commit)}, index.WithExact(name))
+		if err != nil {
+			return nil, err
+		}
+		return &fileset.IndexFilter{
+			F: func(idx *index.Index) bool {
+				return idx.Path == name
+			},
+			R: mr,
+		}, nil
+	}
+	// children
+	getChildren := func() (fileset.FileSource, error) {
+		mr, err := d.storage.NewMergeReader(ctx, []string{compactedCommitPath(file.Commit)}, index.WithPrefix(name+"/"))
+		if err != nil {
+			return nil, err
+		}
+		return &fileset.IndexFilter{
+			F: func(idx *index.Index) bool {
+				return idx.Path != name+"/"
+			},
+			R: mr,
+		}, nil
+	}
+	// create readers
+	exact := NewReader(file.Commit, getExact)
+	children := NewReader(file.Commit, getChildren)
+	// iterate
+	if err := exact.Iterate(ctx, func(finfo *pfs.FileInfoV2, _ fileset.File) error {
+		return cb(finfo)
+	}); err != nil {
+		return err
 	}
-	return glob[:idx[0]]
+	if err := children.Iterate(ctx, func(finfo *pfs.FileInfoV2, _ fileset.File) error {
+		if pathIsChild(name, finfo.File.Path) {
```

Comment:
Why is this necessary? Shouldn't the index filter already handle this?

Reply:
ListFile only returns immediate children, but we have to include recursive children to compute the hashes.

Reply:
Ahh, I see. This handles the skipping of children of children.